### PR TITLE
Rework for introducing changes from the Issue's list

### PR DESCRIPTION
**Pull request by teiclap**
I tried to cover all current Issues.
Most of the Issues in the list related to encryption have been already fixed by John and committed.
Note: a new markdown file "Adaptation_Layer_Indication.md" has been prepared for IANA proposal, please Michael take care of it.

### CODE REVIEW

File: draft-westerlund-tsvwg-dtls-over-sctp-bis.md [hunk 1]
```diff
@@ -314,11 +316,16 @@ TLS:  Transport Layer Security
...
 
-  If an implementation supports partial delivery in both the SCTP API
-  and the ULP API, and also partial processing in the DTLS/SCTP
-  implementation, then the buffering space in the DTLS/SCTP layer
-  ought to be no more than two DTLS records. In which case the MMS to
-  set is dependent on the ULP and the endpoints capabilities.
+   If an implementation supports partial delivery in both the SCTP API
+   and the ULP API, and also partial processing in the DTLS/SCTP
+   implementation, then the buffering space in the DTLS/SCTP layer
+   ought to be no more than two DTLS records. In which case the MMS to
```
Comment (tuexen) [inline draft-westerlund-tsvwg-dtls-over-sctp-bis.md:322]: Can you elaborate on why two DTLS records is enough?

Can't you have on each incoming stream an incomplete DTLS record you can't process at the DTLS layer?

File: draft-westerlund-tsvwg-dtls-over-sctp-bis.md [hunk 2]
```diff
@@ -616,8 +622,7 @@ During resumption, the maximum message size is renegotiated.
    DTLS/SCTP Mandatory Options, in that case the Association will
    proceed as specified in the previous {{DTLS-init}} section.  If the
    peer replies with an INIT-ACK not containing all DTLS/SCTP
-   Mandatory Options, the Client can decide to keep on working with
-   RFC 6083 fallback, plain data only, or to ABORT the association.
+   Mandatory Options, the Client SHOULD reply with ABORT.
```
Comment (tuexen) [inline draft-westerlund-tsvwg-dtls-over-sctp-bis.md:625]: This is hard to realise... At least with SCTP implementations using the socket API specified in [RFC 6458](https://datatracker.ietf.org/doc/html/rfc6458). The normal procedure is that the upper layer enables some options, sets some parameters, then the association is set up and features are negotiated. After the association has been established, the end-points can check and abort the association.

The constraints are coming from the upper layer, so the upper layer should check it. In this case, the upper layer is the DTLS implementation.

File: draft-westerlund-tsvwg-dtls-over-sctp-bis.md [hunk 3]
```diff
@@ -627,14 +632,13 @@ During resumption, the maximum message size is renegotiated.
    follow the sequence for DTLS initialization {{DTLS-init}} and the
    related traffic case.  If a SCTP Server supports DTLS, when
    receiving an INIT chunk with not all DTLS/SCTP Mandatory
-   Options, it can decide to continue by creating an Association with
-   RFC 6083 fallback, plain data only or to ABORT it.
+   Options, it SHOULD reply with ABORT.
```
Comment (tuexen) [inline draft-westerlund-tsvwg-dtls-over-sctp-bis.md:635]: See my comment above. Isn't checking the requirements of an upper layer the job of the upper layer?

**Comment by teiclap** [comment]
Hi Michael,
isn’t the DTLS implementation part of the rfc6083 layer?
The user doesn’t see DTLS actually, or I am wrong?
 
Regards,
Claudio.

**Comment by tuexen** [comment]
> Hi Michael, isn’t the DTLS implementation part of the rfc6083 layer? The user doesn’t see DTLS actually, or I am wrong? 
> Regards, Claudio.

In my view [RFC 6083](https://datatracker.ietf.org/doc/html/rfc6083) describes how to use DTLS over SCTP. It is not describing a single layer, but how to combine two. The application is interacting with the upper layer interface of DTLS.

I'm not sure why you are asking the above question. Can elaborate?

**Comment by teiclap** [comment]
Hi Michael,
in my view rfc6083 is a layer on its own, it offers the same interface as SCTP and has the possibility
to exploit DTLS as an encryption engine.
rfc6083 layer does fragmentation and reassembly of user data and interfaces both DTLS and SCTP.
In my view, DTLS is not even visible to the user.
In the implementation, it may be hidden inside SCTP API (by adding some extra parameters).
 
Regards,
Claudio.

File: draft-westerlund-tsvwg-dtls-over-sctp-bis.md (hunk 1 above)
Comment (gloinul) [reply draft-westerlund-tsvwg-dtls-over-sctp-bis.md:322]: So the comment on two DTLS records are per user message which properly should be made explicit. So you may have multiple of these, so per stream for in order delivery, and for streams with out of order delivery it becomes per message. 

It also includes an assumption that may be wrong in some implementation, that the partial delivery will only release data up to the earliest gap in the byte delivery, i.e. the data prior and including the cumulative ACK, but not ranges of data beyond a gap in the user message will be given to the layer above SCTP, and that SCTP will be responsible for holding this buffer.

File: draft-westerlund-tsvwg-dtls-over-sctp-bis.md (hunk 2 above)
Comment (gloinul) [reply draft-westerlund-tsvwg-dtls-over-sctp-bis.md:625]: Sounds like something that can be resolved with an reformulation. I guess the issue here is depending on how one interpret the meaning of "the client". If that is the SCTP layer only, no I can see the validity of your comment under some API, but I think we are talking about the DTLS + SCTP stack combination that needs to do the right thing here.

Comment (tuexen) [reply draft-westerlund-tsvwg-dtls-over-sctp-bis.md:625]: > Sounds like something that can be resolved with an reformulation. I guess the issue here is depending on how one interpret the meaning of "the client". If that is the SCTP layer only, no I can see the validity of your comment under some API, but I think we are talking about the DTLS + SCTP stack combination that needs to do the right thing here.

Yes.

**Comment by tuexen** [comment]
> Hi Michael, in my view rfc6083 is a layer on its own, it offers the same interface as SCTP and has the possibility to exploit DTLS as an encryption engine. rfc6083 layer does fragmentation and reassembly of user data and interfaces both DTLS and SCTP. In my view, DTLS is not even visible to the user. In the implementation, it may be hidden inside SCTP API (by adding some extra parameters). Regards, Claudio. 

Ahh, I see. I do see DTLS/SCTP like I see TLS/TCP. You can implement TLS on top of TCP by using the upper layer API of TCP. For example, OpenSSL does it like this. The same is true for the DTLS/SCTP implementation of OpenSSL and the SCTP of the FreeBSD or Linux kernel stacks. I'm not saying that this is the only way of implementing DTLS/SCTP, but I think it should be at least one possible way of doing so.

**Comment by gloinul** [comment]
The IANA changed has been removed from this PR and instead have been improved in #44.

File: draft-westerlund-tsvwg-dtls-over-sctp-bis.md (hunk 2 above)
Comment (gloinul) [reply draft-westerlund-tsvwg-dtls-over-sctp-bis.md]: I have just committed some changes to these two paragraphs. @tuexen can you please check if this is better or possible suggest rewording?

Comment (tuexen) [reply draft-westerlund-tsvwg-dtls-over-sctp-bis.md]: Will do after we finished the handshake discussion.

**Pull request merged**